subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/OptionalDependenciesHandler.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * This class also makes a special case of the "optional" configuration, for backwards compatibility: the optional configuration used to store optional dependencies. But if we no longer resolve
+ * optional dependencies, then the optional configuration becomes effectively empty. To avoid this, we ignore the state of optional dependencies if they belong to the "optional" configuration.
+ */
+public class OptionalDependenciesHandler {
+    private final OptionalDependenciesState optionalDependencies;
+    private final ImmutableModuleIdentifierFactory moduleIdentifierFactory;
+    private final DependencySubstitutionApplicator dependencySubstitutionApplicator;
+
+    private boolean started;
+    private boolean isOptionalConfiguration;
+    private List<PendingOptionalDependencies> noLongerOptional;
+
+    public OptionalDependenciesHandler(OptionalDependenciesState optionalDependencies,
+                                       ImmutableModuleIdentifierFactory moduleIdentifierFactory,
+                                       DependencySubstitutionApplicator dependencySubstitutionApplicator) {
+
+        this.optionalDependencies = optionalDependencies;
+        this.moduleIdentifierFactory = moduleIdentifierFactory;
+        this.dependencySubstitutionApplicator = dependencySubstitutionApplicator;
+    }
+
+    private ModuleIdentifier lookupModuleIdentifier(DependencyState dependencyState) {
+        DependencySubstitutionApplicator.SubstitutionResult application = dependencySubstitutionApplicator.apply(dependencyState.getDependencyMetadata());
+        DependencySubstitutionInternal details = application.getResult();
+        if (details != null && details.isUpdated()) {
+            ComponentSelector target = details.getTarget();
+            if (target instanceof ModuleComponentSelector) {
+                ModuleComponentSelector selector = (ModuleComponentSelector) target;
+                return moduleIdentifierFactory.module(selector.getGroup(), selector.getModule());
+            }
+        }
+        return dependencyState.getModuleIdentifier();
+    }
+
+    boolean maybeAddAsOptionalDependency(NodeState node, DependencyState dependencyState) {
+        ModuleIdentifier key = lookupModuleIdentifier(dependencyState);
+        PendingOptionalDependencies pendingOptionalDependencies = optionalDependencies.getPendingOptionalDependencies(key);
+        boolean pending = pendingOptionalDependencies.isPending();
+
+        if (dependencyState.getDependencyMetadata().isOptional() && !isOptionalConfiguration && pending) {
+                pendingOptionalDependencies.addNode(node);
+                return true;
+        }
+        if (pending) {
+            if (noLongerOptional == null) {
+                noLongerOptional = Lists.newLinkedList();
+            }
+            noLongerOptional.add(pendingOptionalDependencies);
```

Comment:
⭕️ This could also be maintained in the state object, by the `noLongerOptional` call.

Reply:
Not really: this one is scoped to a _start/complete_ session, whereas the state object is global to the whole resolution.